### ISSUE TITLE
Support for automatic ECR login refresh?

### ISSUE THREAD

**Issue by d33tah** [pyouroboros/ouroboros#303]
Hi,

I'm considering migrating away from Watchtower to your project. The problem we have is that with watchtower + AWS Elastic Container Registry (ECR), we had to:

1. share /root/.docker/config as a volume,
2. set up a cron job to do the following every few hours (otherwise creds would expire): ```0 */6 * * * /usr/local/bin/aws ecr get-login --no-include-email | bash```

It feels flakey to rely on Cron on host container and we'd prefer to get rid of this dependency. Any chance Ouroboros already supports this use case? We can't just pass REPO_PASS because those credentials regularily change and it would be better if we could pass AWS credentials and have Ouroboros take care of authentication by itself. If it's not supported, would that be a major change to the codebase?

**Comment by DirtyCajunRice** [comment]
We have had a small handful of people that have came to the same conclusion as you. After some digging it looks like if the credentials are successful docker.py loads the credentials into ram and ignores the actual file regardless of the “reading file” that its debug output suggests. It wouldnt be a major change to the codebase and i completely understand the need for revolving passwords. 

My first thought would be to add a param/env var that is similar to the cron/interval but specifically for forcing a “re-auth” from the file on disk. this would still require some automation outside of the container for getting the new credentials for the config file as i dont see a sustainable way to support pulling auth from a variable number of registries, but would allow ouro to continue without a restart. 

Thoughts? Suggestions?

**Comment by d33tah** [comment]
@DirtyCajunRice Thanks for the reply! Having Ouroboros re-read the creds from config would put it on par with Watchtower for this use case, but that's not exactly what I was hoping for. While we're at it, why doesn't it actually just read the credentials every single time?

Ideally, I would like Ouroboros to have some ECR-specific code that would use the AWS credentials to regenerate the credentials without external interaction. Is that within the scope of the project/

You wrote:

>  i dont see a sustainable way to support pulling auth from a variable number of registries

Each registry is tied to its own domain so I would imagine it's easy to handle ECR specially. Or am I wrong?

**Comment by DirtyCajunRice** [comment]
Its just how docker.py was written. They have the .reauth() workaround but i dont think they want that to be loop called. Might just be preference. I have none and would see no issue adding it. 

As for what you quoted, what i meant is that supporting ecr means supporting any/all other platforms that are similar as well as adding boto3 python module (assuming it can do the auth). So im just hesitant to open that can of worms. We can wait for a bit of feedback from people though! Ill try to get to the prior asap.

**Comment by ljconde** [comment]
We're having this same issue with Ouroboros. Actually switched back to Watchtower while it's fixed, because having to re-deploy every 12 hours so that the auth tokens are updated kind of defeats the purpose of auto-updating. In our case, we're using an EC2 attached policy for authentication to our ECR registry, and we refresh auth tokens with Cron before they expire (12h).. but the new auth tokens aren't re-read by Ouroboros, so updating fails because it can't connect to the registry. Is there any workaround to this?

**Comment by ljconde** [comment]
@DirtyCajunRice Would having Ouroboros re-read the credentials from config be a big ask? I'm really only waiting on this to make the switch. I tried restarting the Ouroboros container in hopes that would force the re-read, but that doesn't seem to work. Any ideas on a workaround? Thanks!

**Comment by DirtyCajunRice** [comment]
No, not at all. Ive been very very busy and just havent had any time to pour into ouro for the next wave of updates. Please bare with me! If you are comfortable with python and can do it sooner feel free to make a PR and I can chat through it with you

**Comment by DirtyCajunRice** [comment]
>No, not at all. Ive been very very busy and just havent had any time to pour into ouro for the next wave of updates. Please bare with me! If you are comfortable with python and can do it sooner feel free to make a PR and I can chat through it with you.

restarting the container should read from disk again unless you are mounting the file instead of the folder though

**Comment by ljconde** [comment]
@DirtyCajunRice I tried mounting the folder instead of the file and I'm getting the same result. I replaced `/root/.docker/config.json` with just `/root/.docker`, but that didn't do the trick. Here's what I have at the moment:
`  ouroboros:
    image: pyouroboros/ouroboros
    environment:
      - CLEANUP=true
      - INTERVAL=30
      - LOG_LEVEL=info
      - SELF_UPDATE=true
      - LABEL_ENABLE=true
      - LABEL_ONLY=true
      - NOTIFIERS="slack://mavenaudience-prod@xxx/yyy/zzz"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - /root/.docker:/root/.docker
`
If you point me in the right direction I can work on a PR that includes the change! Thanks!

**Comment by DirtyCajunRice** [comment]
I only meant that if you mount the folder vs the file a _restart_ would read from file.
As for direction pointing:
https://github.com/pyouroboros/ouroboros/blob/646cb95527e8a54ee5040b2c9a7395cf6d98b198/pyouroboros/dockerclient.py#L14 is how it connects and it only does on init.

the hacky and fast way to do it would probably be just to add the connect here as well below this line
https://github.com/pyouroboros/ouroboros/blob/646cb95527e8a54ee5040b2c9a7395cf6d98b198/pyouroboros/dockerclient.py#L410
which would make it "connect" every time it did an update in addition to init

**Comment by ljconde** [comment]
Yeah, had the same result after container restart, which I can't explain really.. unless docker is still caching the file because it's still the only file in that folder? Not sure. I'll ticker around the code to test out what you're proposing and let you know! Thanks.

**Comment by mogwai** [comment]
I've solved the problem by creating a cron job that logs in and restarts the container using the following script:

crontab -e

```sh
@hourly  /root/login.sh
```

```sh
#!/bin/bash

$(aws ecr get-login --no-include-email)
cd /root/my-project
docker-compose restart image-updater
```
